File: android/src/main/java/com/rnstartiosdk/RNStartIoNativeAd.kt
package com.rnstartiosdk

import android.util.Log
import android.view.View
import com.facebook.react.uimanager.ThemedReactContext
import com.margelo.nitro.rnstartiosdk.CampaignAction
import com.margelo.nitro.rnstartiosdk.HybridRNStartIoNativeAdSpec
import com.margelo.nitro.rnstartiosdk.NativeAdDetails
import com.startapp.sdk.ads.nativead.StartAppNativeAd
import com.startapp.sdk.adsbase.Ad
import com.startapp.sdk.adsbase.adlisteners.AdEventListener

class RNStartIoNativeAd(val context: ThemedReactContext): HybridRNStartIoNativeAdSpec() {
    private companion object {
        val LOG_TAG: String = RNStartIoNativeAd::class.java.simpleName
    }

    override val view: View = View(context)
    override var onLoadAd: (NativeAdDetails) -> Unit = {}
    override var onLoadError: ((String) -> Unit)? = {}

    init {
        val nativeAd = StartAppNativeAd(context)
        nativeAd.loadAd(object : AdEventListener {
            override fun onReceiveAd(ad: Ad) {
                Log.v(LOG_TAG, "loadNative: onReceiveAd")

                val nativeAds = nativeAd.nativeAds
                if (nativeAds != null && nativeAds.isNotEmpty()) {
                    val nativeAdDetails = nativeAds[0]
                    if (nativeAdDetails != null) {
                        nativeAdDetails.registerViewForInteraction(view)
                        onLoadAd(
                            NativeAdDetails(
                                nativeAdDetails.title,
                                nativeAdDetails.description,
                                nativeAdDetails.rating.toDouble(),
                                nativeAdDetails.imageUrl,
                                nativeAdDetails.secondaryImageUrl,
                                nativeAdDetails.installs,
                                nativeAdDetails.category,
                                nativeAdDetails.packageName,
                                if (nativeAdDetails.campaignAction.name == CampaignAction.LAUNCH_APP.name)
                                    CampaignAction.LAUNCH_APP
                                else CampaignAction.OPEN_MARKET,
                                nativeAdDetails.callToAction
                            )
                        )
                    } else {
                        onLoadError?.invoke("No ads available")
                    }
                } else {
                    onLoadError?.invoke("No ads available")
                }
            }

            override fun onFailedToReceiveAd(ad: Ad?) {
                Log.v(LOG_TAG, "loadNative: onFailedToReceiveAd: " + (ad?.errorMessage))
                onLoadError?.invoke(ad?.errorMessage ?: "Failed to load error")
            }
        })
    }
}

File: ios/RNStartIoNativeAd.swift
//
//  RNStartIoNativeAd.swift
//  RNStartIoSdk
//
//  Created by Abhijit Dey on 08/09/25.
//

import Foundation
import StartApp
import NitroModules
import UIKit

class RNStartIoNativeAd: HybridRNStartIoNativeAdSpec {
    
    private var isNativeAdLoading = false
    private var isNativeAdLoaded = false
    var currentLoadShowDelegate: LoadShowAdDelegate?

    var onLoadAd: (NativeAdDetails) -> Void = { _ in }
    var onLoadError: ((String) -> Void)? = nil
    
    var view: UIView = CustomView()
    
    func afterUpdate() {
        requestNativeAd()
    }
    
    private func requestNativeAd(){
        guard !isNativeAdLoading, !isNativeAdLoaded else { return }
        isNativeAdLoading = true
        
        let loadShowAdDelegate = LoadShowAdDelegate()
        let startAppNativeAd = STAStartAppNativeAd()
        currentLoadShowDelegate = loadShowAdDelegate
        
        loadShowAdDelegate.loadPromise.then({
            self.isNativeAdLoaded = true
            guard let adDetails = startAppNativeAd.adsDetails,
                  !(adDetails.count < 1),
                  let nativeAdDetails = adDetails[0] as? STANativeAdDetails else {
                self.onLoadError?("No ads available")
                self.isNativeAdLoading = false
                return
            }
            (self.view as? CustomView)?.didAppear {
                nativeAdDetails.registerView(forImpressionAndClick: self.view)
            }
            self.onLoadAd(
                NativeAdDetails(
                    title: nativeAdDetails.title ?? "",
                    description: nativeAdDetails.description ?? "",
                    rating: nativeAdDetails.rating != nil ? nativeAdDetails.rating.doubleValue : 0,
                    imageUrl: nativeAdDetails.imageUrl,
                    secondaryImageUrl: nativeAdDetails.secondaryImageUrl,
                    installs: "",
                    category: nativeAdDetails.category ?? "",
                    packageName: "",
                    campaignAction: nativeAdDetails.clickToInstall == "install" ?
                                        CampaignAction.openMarket :
                                        CampaignAction.launchApp,
                    callToAction: nativeAdDetails.callToAction ?? ""
                )
            )
            self.isNativeAdLoading = false
        }).catch({error in
            self.onLoadError?(error.localizedDescription)
            self.isNativeAdLoading = true
        })
        
        startAppNativeAd.load(withDelegate: loadShowAdDelegate)
    }
    
}
